pkg/queue/breaker.go

```diff
@@ -99,25 +128,55 @@ type Semaphore struct {
 	mux      sync.Mutex
 }
 
-// Acquire receives the token from the semaphore, potentially blocking
+// Acquire receives the token from the semaphore, potentially blocking.
 func (s *Semaphore) Acquire() {
 	<-s.queue
 }
 
-// Release releases the token to the queue
-// The operation is potentially blocking when the queue is full
+// Release potentially puts the token back to the queue.
+// If the semaphore capacity was reduced in between and is not yet reflected,
+// we remove the tokens from the rotation instead of returning them back.
 func (s *Semaphore) Release() {
-	s.AddCapacity(1)
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	if s.reducers > 0 {
+		s.capacity -= s.reducers
+		s.reducers--
+	} else {
+		// We want to make sure releasing a token is always non-blocking.
+		select {
+		case s.queue <- s.token:
+		default:
+		}
+	}
+}
+
+// AddCapacity increases the number of tokens in the rotation.
+// Increases only if the size is positive, does nothing otherwise.
+// An error is returned if not all capacity could be added to the semaphore.
+// This error could happen when the number of tokens exceeds the semaphore's queue size.
+func (s *Semaphore) AddCapacity(size int32) error {
+	if size > 0 {
+		for i := int32(0); i < size; i++ {
+			select {
+			case s.queue <- s.token:
+				s.capacity++
```

Comment:
Don't you need to obtain the Lock to make this change?

Does this negate the results of "Split semaphore's capacity operations from acquiring/releasing the token."?

Reply:
Yes, added it. 

No, we still need to split the semaphore's get/put operations from add/reduce capacity to keep the capacity bookkeeping right. Since capacity is the number of tokens in the rotation we don't want them to increase once the tokens are returned.

Reply:
So does this redefine what "Capacity" is? Is capacity the leftover slots available or is it the overall amount of slots (free + taken)?

Reply:
I  see, so it's the theoretical limit.